bip-0360.mediawiki

```diff
@@ -167,389 +171,304 @@ HASH160 <ref name="hash160">Used by P2PKH, P2SH, and P2WPKH addresses, though no
+* Spend requires a SLH-DSA signature
+
+In the event that Schnorr signatures are broken, users can spend their coins using ML-DSA. 
+If both Schnorr and ML-DSA are broken, the user can still rely on SLH-DSA.
+
+One intent in supporting Schnorr, ML-DSA, and SLH-DSA in tapscript, is to allow parties to construct outputs such that funds 
+are still secure even if two of the three the signature algorithms are completely broken. This is motivated by the use case
+of securely storing Bitcoins in a cold wallet for very long periods of time (50 to 100 years).
+
+For PQ signatures we considered the NIST approved SLH-DSA (SPHINCS+), ML-DSA (CRYSTALS-Dilithium), 
+FN-DSA (FALCON). Of these three algorithms, SLH-DSA has the largest signature size, but is the most conservative 
+choice from a security perspective because SLH-DSA is based on well studied and time-tested hash-based cryptography. 
+Both FN-DSA and ML-DSA signatures are significantly smaller than SLH-DSA signatures but are based on newer lattice-based 
+cryptography. Since ML-DSA and FN-DSA are both similar lattice-based designs, we choose to only support one of them as the 
+additional value in diversity of cryptographic assumptions would be marginal. It should be noted that ML-DSA and FN-DSA do
+rely on different lattice assumptions and it may be that case that a break in one algorithm's assumptions would not necessarily
+break the assumptions used by the other other algorithm.
+
+We also considered SQIsign. While it outperforms the three other PQ signature algorithms by having the smallest signatures,
+it has the worst verification performance and requires a much more complex implementation. We may revisit SQIsign separately in the
+future as recent research shows massive performance improvements to SQIsign in version 2.0.<ref name="sqisign2"> "[SQIsign] signing is now nearly 20× faster, at 103.0 Mcycles, and verification is more than 6× faster, at 5.1 Mcycles" [https://csrc.nist.gov/csrc/media/Projects/pqc-dig-sig/documents/round-2/spec-files/sqisign-spec-round2-web.pdf SQIsign: Algorithm specifications and supporting documentation Version 2.0 (February 5 2025)]</ref>.
+
+ML-DSA is intended as the main PQ signature algorithm in Bitcoin. It provides a good balance of security, performance 
+and signature size and is likely to be the most widely supported PQ signature algorithm on the internet. SLH-DSA has a radically 
+different design and set of cryptographic assumptions than ML-DSA. As such SLH-DSA provides an effective
+hedge against an unexpected cryptanalytic breakthrough.
+
+P2QRH, ML-DSA, and SLH-DSA could be activated simultaneously in a single soft fork or P2QRH could be activated first and then
+ML-DSA and SLH-DSA could be independently activated. If at some future point another signature
+algorithm was desired it could follow this pattern.
+
+We consider two different paths for activating PQ signatures in Bitcoin. The first approach is to redefine OP_SUCCESSx opcodes for each
```

Comment:
`OP_CHECKSIG`, `OP_CHECKSIGVERIFY`, and `OP_CHECKSIGADD` all can support other signing algorithms without new opcodes. If the public key is *not* exactly 32 bytes long (x-only public key), then the signature check is considered to have succeeded. Of course there are other considerations, and I don't know if the proposed PQ schemes can be fit into these opcodes (especially given the stack size limit).

See https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki#rules-for-signature-opcodes

Reply:
That's interesting. I didn't know that

SLH-DSA has 32 byte public keys, ML-DSA has much larger public keys but we use a public key hash in the tapleaf so it is also 32 bytes. In the tapleaf world we would probably want to make the public keys 33-bytes with one byte to flag the type of algorithm. It's cool that we could softfork that in. I think we would still need the new tapleaf version for the increased initial stack element size if we were doing OP_SUCCESSx opcodes.